codegen/src/main/java/io/improbable/keanu/codegen/python/VertexProcessor.java

```diff
@@ -114,15 +116,37 @@ private static String toCastedPythonParams(String[] pythonParameters, Class<?>[]
         return String.join(", ", pythonParams);
     }
 
-    private static String toCastedPythonParam(String pythonParameter, Class<?> parameterType) {
+    private static String toCastedPythonParam(String pythonParameter, Class<?> parameterClass) {
+        Class parameterType = Primitives.wrap(parameterClass);
+
         if (DoubleVertex.class.isAssignableFrom(parameterType)) {
-            return "cast_to_double(" + pythonParameter + ")";
+            return "cast_to_double_vertex(" + pythonParameter + ")";
         } else if (IntegerVertex.class.isAssignableFrom(parameterType)) {
-            return "cast_to_integer(" + pythonParameter + ")";
+            return "cast_to_integer_vertex(" + pythonParameter + ")";
         } else if (BooleanVertex.class.isAssignableFrom(parameterType)) {
-            return "cast_to_bool(" + pythonParameter + ")";
+            return "cast_to_boolean_vertex(" + pythonParameter + ")";
+        } else if (Vertex.class.isAssignableFrom(parameterType)) {
+            return "cast_to_vertex(" + pythonParameter + ")";
+        } else if (DoubleTensor.class.isAssignableFrom(parameterType)) {
+            return "cast_to_double_tensor(" + pythonParameter + ")";
+        } else if (IntegerTensor.class.isAssignableFrom(parameterType)) {
+            return "cast_to_integer_tensor(" + pythonParameter + ")";
+        } else if (BooleanTensor.class.isAssignableFrom(parameterType)) {
+            return "cast_to_boolean_tensor(" + pythonParameter + ")";
+        } else if (Double.class.isAssignableFrom(parameterType)) {
+            return "cast_to_double(" + pythonParameter + ")";
+        } else if (Integer.class.isAssignableFrom(parameterType) || Long.class.isAssignableFrom(parameterType)) {
+            return "cast_to_integer(" + pythonParameter + ")";
+        } else if (String.class.isAssignableFrom(parameterType)) {
+            return "cast_to_string(" + pythonParameter + ")";
+        } else if (Long[].class.isAssignableFrom(parameterType) || long[].class.isAssignableFrom(parameterType)) {
+            return "cast_to_long_array(" + pythonParameter + ")";
+        } else if (Integer[].class.isAssignableFrom(parameterType) || int[].class.isAssignableFrom(parameterType)) {
+            return "cast_to_int_array(" + pythonParameter + ")";
+        } else if (Vertex[].class.isAssignableFrom(parameterType)) {
+            return "cast_to_vertex_array(" + pythonParameter + ")";
         } else {
-            return pythonParameter;
+            throw new IllegalArgumentException("Failed to Encode " + pythonParameter + " of type: " + parameterType);
```

Comment:
As discussed - there's a class hierarchy thing going on at the moment so we can't just use a plain old map without us having an unwieldy maintenance job of putting every vertex in there.  We could write a smarter walker that checks the map - but I'd have to convince myself that would definitely work in all cases before I'd want to make the change (and it would potentially reduce the readability a bit anyway).  I'll leave as is for now, and have a think about how we could better handle this (since we have similar code in the Protobuf stuff too).